### ISSUE TITLE
Suggestion: Make DataFrames more useful

### ISSUE THREAD

**Issue by BenjaminBossan** [skorch-dev/skorch#501]
At the moment, passing a pandas DataFrame to `net.fit` seems like it would rarely be useful. It is treated similarly to a dictionary because each column is treated as key and the corresponding value as a 1-dim array, but dictionaries can also hold more dimensional data. I suspect that nobody uses DataFrames with skorch.

I would propose not to change this, but to add a useful transformer (inspired by [fastai's TabularDataset](https://docs.fast.ai/tabular.data.html#TabularDataset) that converts a DataFrame into a dictionary. This transformer could be used in a pipeline as a preprocessing step. It would do the following:

* float dtypes: concatenate to a single array with a specific key (e.g. `X`)
* categorical dtypes: leave each as separate values, with the key corresponding to the column name
* integer dtypes: either treat as cardinal (like float) or as categorical, depending on a parameter
* other dtypes: not sure, maybe raise an error

Using this transformer, the module could pass the cardinal features to a dense layer and the categorical features to an embedding layer (or one-hot encode).

The code could look somewhat like this:

```python
df = pd.DataFrame({
    'rooms': [2, 4, 3, ...],
    'area': [103.5, 42., 234.5, ...],
    'state': ['CA', 'FL', 'NY', ...],
})
df['state'] = df['state'].astype('category')

class MyModule(nn.Module):
    ...
    def forward(self, X, state):
        # X has shape (n, 2) and contains rooms, area
        # state has shape (n,) and contains ints
        X_state = self.embed(state)
        X = torch.cat((X, X_state), axis=0)
        ...

model = Pipeline([
    ('preproc', DataFrameTransformer(treat_int_as_categorical=False)),
    ('net', net),
])
```

My suggestion would be to add this transformer to `helper.py` and to not change any of the existing behavior. The transformer could also have a method that helps with building the right module by giving information about the correct keys and sizes.

Do you think this would be useful and fit the scope of skorch?

**Comment by thomasjpfan** [comment]
I think this is pretty useful. For their [example with rossmann](https://github.com/fastai/course-v3/blob/master/nbs/dl1/lesson6-rossmann.ipynb), they heavy use `Embedding` for their categories.

On a ML level, it would be interesting to see how it compares to one-hot encoding.

To simplify the API, I do not think we need `treat_int_as_categorical`. If a user wants to have categories, they should convert the dtype of that column to a pandas category.

**Comment by BenjaminBossan** [comment]
> On a ML level, it would be interesting to see how it compares to one-hot encoding.

I thought about whether the transformer should support one-hot encoding. But I think it would make the API really messy if we did that. Instead, the user could initialize an identity matrix as embedding matrix (without gradient) to achieve the same effect. 

> To simplify the API, I do not think we need `treat_int_as_categorical`

I believe having this would be really convenient. Also, it is quite unclear what the "correct" behavior should be for integer types.

**Comment by thomasjpfan** [comment]
Sorry if I was a little unclear. I do not think we should do one hot encoding in this API. It would be interesting as a research topic to compare using the embedding and comparing it to traditional categorical encodings.

As for ints, treat as float is a reasonable and expected default. If a user wants it to be treated as a category, we can direct uses to “astype(“category”)” on the int.

**Comment by BenjaminBossan** [comment]
Fixed by #507
